pkg/fqdn/cache.go

```diff
@@ -640,3 +640,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
 
 	return nil
 }
+
+// DNSPendingDelete is an IP that has expired or been evicted from a DNS cache.
+// It records the DNS name and IP, along with other bookkeeping timestamps that
+// help determine when it can be finally deleted. Entries are deletable when
+// they are not marked active by CT GC.
+type DNSPendingDelete struct {
+	Names           []string  `json:"names,omitempty"`
+	IP              net.IP    `json:"ip,omitempty"`
+	ActiveAt        time.Time `json:"active-at,omitempty"`
+	DeletePendingAt time.Time `json:"delete-pending-at,omitempty"`
+}
+
+// DNSPendingDeletes collects DNS Name->IP mappings that may be deletable. They
+// are marked active, periodically, by the CT GC goroutine. When .GC is called,
+// active and deletable entries are returned, allowing us to skip deleting an IP
+// from the global DNS cache to avoid breaking connections that outlast the DNS
+// TTL.
+type DNSPendingDeletes struct {
+	lock.Mutex
+	deletes        map[string]*DNSPendingDelete // map[ip]toDelete
+	lastCTGCUpdate time.Time
+}
+
+// NewDNSPendingDeletes constructs a DNSPendingDeletes that is read to use
+func NewDNSPendingDeletes() *DNSPendingDeletes {
+	return &DNSPendingDeletes{
+		deletes: make(map[string]*DNSPendingDelete),
+	}
+}
+
+// WantToDelete enqueues the ip -> qname as a possible deletion
+// updatedExisting is true when an earlier enqueue existed and was updated
+func (d *DNSPendingDeletes) WantToDelete(ipStr string, qname ...string) (updatedExisting bool) {
+	d.Lock()
+	defer d.Unlock()
+
+	entry, updatedExisting := d.deletes[ipStr]
+	if !updatedExisting {
+		entry = &DNSPendingDelete{}
+		d.deletes[ipStr] = entry
+	}
+
+	entry.Names = append(entry.Names, qname...)
+	entry.IP = net.ParseIP(ipStr)
+	entry.DeletePendingAt = time.Now()
+
+	return updatedExisting
+}
+
+// Delete after it is seen by CT GC, DeletePendingAt < lastCTGCUpdate, and the ActiveTime is 0 or no longer current, ActiveAt < lastCTGCUpdate
+func (d *DNSPendingDeletes) canDelete(del *DNSPendingDelete) bool {
```

Comment:
I'm having troubles understanding this function.
```
// Delete after it is seen by CT GC, DeletePendingAt < lastCTGCUpdate, and the ActiveTime is 0 or no longer current, ActiveAt < lastCTGCUpdate
```

> // Delete after it is seen by CT GC

which logic is doing this?

>  DeletePendingAt < lastCTGCUpdate

go it, is `d.lastCTGCUpdate.After(del.DeletePendingAt)`

> ActiveTime is 0 or no longer current

what's "current"?

> ActiveAt < lastCTGCUpdate

go it, is `d.lastCTGCUpdate.After(del.ActiveAt)`

Perhaps the part I'm having more trouble understanding is the field names. To me it looks like we can delete if the last CT GC update was executed after the time in the field "delete pending at" which is fine. Now for the second part, we can only delete if the last CT GC update was executed after the entry was active or 0? What does it mean an entry to be "active"?

Reply:
This function took me a while to get right the first time... and when I switched it to isAlive I ended up messing it up again!

One note on the new version of this (isAlive): there aren't any `time.BeforeOrEqual` type functions so I actually kept the older logic and added a not to the start :/

>>// Delete after it is seen by CT GC

> which logic is doing this?

The logic for the CT GC is in `pkg/maps/ctmap/gc/gc.go`. I can state that directly here if you want (but it is mentioned in the Endpoint function that calls this function).

>> ActiveTime is 0 or no longer current

> what's "current"?

"current" was a bad word. In the context of the CT GC cycle, current meant that something was active/alive during the most recent CT GC cycle. The comment doesn't use current anymore, but it may still not be clear.

Basically we can only delete an entry (now called a zombie) if a CT GC cycle completed and it was not marked active/alive. The code for this would look like `lastCTGCUpdate.After(DeletePendingAt) && lastCTGCUpdate.After(ActiveAt)` but, we need to guard against `ActiveAt` not being set, because only the CT GC can set it but the DNS garbage collector can check if entries/zombies are alive/active. Since DeletePendingAt is always set, 
In the new code, this function is called `isAlive` but it's actually easier to keep the canDelete logic and invert it. The alternative statement would be much larger, and we would need to check `ActiveAt.IsZero()`. It would be something like:
`(DeletePendingAt.After(lastGCUpdate) || DeletePendingAt.Equal(lastGCUpdate)) ||
  (AliveAt.After(lastGCUpdate) || AliveAt.Equal(lastGCUpdate))`